README.md

# react-native-audio-streaming

## Features

- Background audio streaming of remote stream
- Control via sticky notification on android and media center on iOS
- Shoutcast/Icy meta data support
- Simple UI player component (if needed, an api to control the sound is available)

If you are only looking to play local audio file with app in foreground, please [see other audio libs](https://github.com/tlenclos/react-native-audio-streaming/blob/master/README.md#other-rn-audio-projects).

![Demo iOS](https://raw.githubusercontent.com/tlenclos/react-native-audio-streaming/master/demo_ios.gif)
![Demo android](https://raw.githubusercontent.com/tlenclos/react-native-audio-streaming/master/demo_android.gif)

## First installation step (applied for both iOS & Android)

`$ npm install react-native-audio-streaming --save`

### Next installation steps for iOS (choose one of three options below)

#### 1. Cocoapods installation

1. add `pod 'RNAudioStreaming', :path => '../node_modules/react-native-audio-streaming'` to `Podfile`
2. run `pod install`

#### 2. Mostly automatic installation

`$ react-native link react-native-audio-streaming`

Go to `node_modules` ➜ `react-native-audio-streaming` => `ios` => `Pods` and drag/drop `Pods.xcodeproj` to the Libraries folder in your XCode project.

In XCode, in the project navigator, select your project. Add `libReactNativeAudioStreaming.a` and `libStreamingKit.a` to your project's `Build Phases` ➜ `Link Binary With Libraries`

#### 3. Manual installation

1. In XCode, in the project navigator, right click `Libraries` ➜ `Add Files to [your project's name]`
2. Go to `node_modules` ➜ `react-native-audio-streaming` => `ios`
   - run `pod install` to download StreamingKit dependency
   - add `ReactNativeAudioStreaming.xcodeproj` to the Libraries folder in your XCode project
   - add `Pods/Pods.xcodeproj` to the Libraries folder in your XCode project
3. In XCode, in the project navigator, select your project. Add `libReactNativeAudioStreaming.a` and `libStreamingKit.a` to your project's `Build Phases` ➜ `Link Binary With Libraries`
4. Run your project (`Cmd+R`)

### Final steps for iOS (required after doing any of three options above)

1. Make sure `$(SRCROOT)/../node_modules/react-native-audio-streaming/ios` is added to your project's `Header Search Paths` within the `Build Settings` section.

2. Update Info.plist file of your Xcode project and add audio background mode

```xml
    <key>UIBackgroundModes</key>
    <array>
      <string>audio</string>
    </array>
```

### Next installation steps for Android

1. Open up `android/app/src/main/java/[...]/MainApplication.java`
  - Add `import com.audioStreaming.ReactNativeAudioStreamingPackage;` to the imports at the top of the file
  - Add `new ReactNativeAudioStreamingPackage()` to the list returned by the `getPackages()` method
  If you're using Android 23 or above
  - Add `new ReactNativeAudioStreamingPackage(MainActivity.class)` to he list returned by the `getPackages()`method instead.
2. Append the following lines to `android/settings.gradle`:
  	```
  	include ':react-native-audio-streaming'
  	project(':react-native-audio-streaming').projectDir = new File(rootProject.projectDir, 	'../node_modules/react-native-audio-streaming/android')
  	```
3. Insert the following lines inside the dependencies block in `android/app/build.gradle`:
  	```
      compile project(':react-native-audio-streaming')
  	```
  	
## Usage

### Playing sound (similar code used by the player UI)

```javascript
import { ReactNativeAudioStreaming } from 'react-native-audio-streaming';

const url = "http://lacavewebradio.chickenkiller.com:8000/stream.mp3";
ReactNativeAudioStreaming.pause();
ReactNativeAudioStreaming.resume();
ReactNativeAudioStreaming.play(url, {showIniOSMediaCenter: true, showInAndroidNotifications: true});
ReactNativeAudioStreaming.stop();
```

For more information see the Example app.

### Player UI

```javascript
import { Player } from 'react-native-audio-streaming';

class PlayerUI extends Component {
  render() {
    return (
        <Player url={"http://lacavewebradio.chickenkiller.com:8000/stream.mp3"} />
    );
  }
}
```

## TODO

- [ ] Allow to play local files
- [ ] Allow to specify custom style for the android notification (maybe a custom view ?)
- [ ] Allow to specify custom styles for the player
- [ ] Handle artwork of artist
- [ ] Add tests

## Other RN audio projects

- [jsierles/react-native-audio](https://github.com/jsierles/react-native-audio) to play local audio and record
- [zmxv/react-native-sound](https://github.com/zmxv/react-native-sound) to play local audio with more controls

## Credits

- Android version based on the work of @EstebanFuentealba https://github.com/EstebanFuentealba/react-native-android-audio-streaming-aac
- iOS version based on the work of @jhabdas https://github.com/jhabdas/lumpen-radio

See also the list of [contributors](https://github.com/tlenclos/react-native-audio-streaming/graphs/contributors) who participated in this project.

## Contribute

Since symlink support is [still lacking](https://github.com/facebook/react-native/issues/637) on React Native, I use the [wml](https://github.com/wix/wml) cli tool created by the nice folks at wix.

`wml add ~/react-native-audio-streaming ~/react-native-audio-streaming/Example/node_modules/react-native-audio-streaming`

## [Changelog](https://github.com/tlenclos/react-native-audio-streaming/blob/master/CHANGELOG.md)

## License

This project is licensed under the MIT License - see the LICENSE file for details


CHANGELOG.md
# Change Log

## [v2.3.2](https://github.com/tlenclos/react-native-audio-streaming/tree/v2.3.2) (2017-02-25)
[Full Changelog](https://github.com/tlenclos/react-native-audio-streaming/compare/v2.3.1...v2.3.2)

**Fixed bugs:**

- Android notifcation player doesn't go away [\#49](https://github.com/tlenclos/react-native-audio-streaming/issues/49)

**Closed issues:**

- Support for RN \>=0.40 [\#72](https://github.com/tlenclos/react-native-audio-streaming/issues/72)
- Cannot read property 'play' of undefined [\#64](https://github.com/tlenclos/react-native-audio-streaming/issues/64)
- react native \>= 0.4 [\#63](https://github.com/tlenclos/react-native-audio-streaming/issues/63)
- RN 0.39.2 Playback error: Wrong \# arguments [\#59](https://github.com/tlenclos/react-native-audio-streaming/issues/59)
- \[Android\] ReactNativeAudioStreamingModule.getClassActivity null pointer exception [\#51](https://github.com/tlenclos/react-native-audio-streaming/issues/51)

**Merged pull requests:**

- Fix for Android player notification not closing on app close [\#74](https://github.com/tlenclos/react-native-audio-streaming/pull/74) ([IndrekV](https://github.com/IndrekV))
- Mention adding path to Header Search Path [\#73](https://github.com/tlenclos/react-native-audio-streaming/pull/73) ([Naoto-Ida](https://github.com/Naoto-Ida))
- Wrong relative path for Pod [\#71](https://github.com/tlenclos/react-native-audio-streaming/pull/71) ([Naoto-Ida](https://github.com/Naoto-Ida))
- Use empty string instead of missing StreamTitle. [\#69](https://github.com/tlenclos/react-native-audio-streaming/pull/69) ([hudashot](https://github.com/hudashot))
- Fix XML in info.plist for iOS xcodeproj [\#68](https://github.com/tlenclos/react-native-audio-streaming/pull/68) ([AdamLiechty](https://github.com/AdamLiechty))
- Missing step to Pods.xcodeproj file [\#62](https://github.com/tlenclos/react-native-audio-streaming/pull/62) ([jsaucedo](https://github.com/jsaucedo))

## [v2.3.1](https://github.com/tlenclos/react-native-audio-streaming/tree/v2.3.1) (2016-12-15)
[Full Changelog](https://github.com/tlenclos/react-native-audio-streaming/compare/v2.3.0...v2.3.1)

**Closed issues:**

- How to debug? [\#47](https://github.com/tlenclos/react-native-audio-streaming/issues/47)
- Error while building [\#46](https://github.com/tlenclos/react-native-audio-streaming/issues/46)
- Not able to install..... [\#43](https://github.com/tlenclos/react-native-audio-streaming/issues/43)

**Merged pull requests:**

- Add Cocoapods RNAudioStreaming.podspec file [\#54](https://github.com/tlenclos/react-native-audio-streaming/pull/54) ([siriusdely](https://github.com/siriusdely))

## [v2.3.0](https://github.com/tlenclos/react-native-audio-streaming/tree/v2.3.0) (2016-11-18)
[Full Changelog](https://github.com/tlenclos/react-native-audio-streaming/compare/v2.2.0...v2.3.0)

**Implemented enhancements:**

- Can we make notification on android and media center control on iOS optional ? [\#27](https://github.com/tlenclos/react-native-audio-streaming/issues/27)

**Closed issues:**

- Stop working when app in background. [\#39](https://github.com/tlenclos/react-native-audio-streaming/issues/39)
- hi! I used Redux,  I do not know how to introduce this [\#38](https://github.com/tlenclos/react-native-audio-streaming/issues/38)
- Android get URL of current track like iOS [\#35](https://github.com/tlenclos/react-native-audio-streaming/issues/35)
- libaac does not work in android Marshmallow & above [\#30](https://github.com/tlenclos/react-native-audio-streaming/issues/30)
- can't get work with rn 0.31.0 [\#25](https://github.com/tlenclos/react-native-audio-streaming/issues/25)
- Unable to run pod install [\#17](https://github.com/tlenclos/react-native-audio-streaming/issues/17)

## [v2.2.0](https://github.com/tlenclos/react-native-audio-streaming/tree/v2.2.0) (2016-10-24)
[Full Changelog](https://github.com/tlenclos/react-native-audio-streaming/compare/v2.1.0...v2.2.0)

**Closed issues:**

- Can't get to work in Android 23 [\#24](https://github.com/tlenclos/react-native-audio-streaming/issues/24)
- Where do I get libStreamingKit.a? [\#20](https://github.com/tlenclos/react-native-audio-streaming/issues/20)
- Where do I get Pods.xcodeproj? [\#19](https://github.com/tlenclos/react-native-audio-streaming/issues/19)
- not able to control play/pause remotely using earbuds [\#18](https://github.com/tlenclos/react-native-audio-streaming/issues/18)
- Not working with react-native-navigation [\#14](https://github.com/tlenclos/react-native-audio-streaming/issues/14)
- iOS build error: STKAudioPlayer.h not found [\#10](https://github.com/tlenclos/react-native-audio-streaming/issues/10)

**Merged pull requests:**

- Update aac libraries for android [\#32](https://github.com/tlenclos/react-native-audio-streaming/pull/32) ([madhu314](https://github.com/madhu314))
- Update README.md [\#31](https://github.com/tlenclos/react-native-audio-streaming/pull/31) ([brettpappas](https://github.com/brettpappas))
- Make notifications on android and media center on iOS optional [\#29](https://github.com/tlenclos/react-native-audio-streaming/pull/29) ([madhu314](https://github.com/madhu314))
- Fix for iOS build error [\#28](https://github.com/tlenclos/react-native-audio-streaming/pull/28) ([brettpappas](https://github.com/brettpappas))
- Update README.md [\#26](https://github.com/tlenclos/react-native-audio-streaming/pull/26) ([donedgardo](https://github.com/donedgardo))
- corrects README to specify correct file for Android install [\#23](https://github.com/tlenclos/react-native-audio-streaming/pull/23) ([dininski](https://github.com/dininski))
- Update react-native to 0.31.0 [\#11](https://github.com/tlenclos/react-native-audio-streaming/pull/11) ([tlenclos](https://github.com/tlenclos))

## [v2.1.0](https://github.com/tlenclos/react-native-audio-streaming/tree/v2.1.0) (2016-09-06)
[Full Changelog](https://github.com/tlenclos/react-native-audio-streaming/compare/v2.0.2...v2.1.0)


Example/App.js

import React, { Component } from 'react';
import {
    AppRegistry,
    StyleSheet,
    Text,
    View,
    ListView,
    Platform,
    TouchableOpacity
} from 'react-native';
import { ReactNativeAudioStreaming, Player } from 'react-native-audio-streaming';

export default class App extends Component {
    constructor() {
        super();
        this.ds = new ListView.DataSource({rowHasChanged: (r1, r2) => r1 !== r2});
        this.urls = [
            {
                name: 'Shoutcast stream',
                url: 'http://lacavewebradio.chickenkiller.com:8000/stream.mp3'
            },
            {
                name: 'M4A stream',
                url: 'http://web.ist.utl.pt/antonio.afonso/www.aadsm.net/libraries/id3/music/02_Viandanze.m4a'
            },
            {
                name: 'MP3 stream with ID3 meta data',
                url: 'http://web.ist.utl.pt/antonio.afonso/www.aadsm.net/libraries/id3/music/Bruno_Walter_-_01_-_Beethoven_Symphony_No_1_Menuetto.mp3'
            },
            {
                name: 'MP3 stream',
                url: 'http://www.stephaniequinn.com/Music/Canon.mp3'
            }
        ];

        this.state = {
            dataSource: this.ds.cloneWithRows(this.urls),
            selectedSource: this.urls[0].url
        };
    }

    render() {
        return (
            <View style={styles.container}>
                <ListView
                    dataSource={this.state.dataSource}
                    renderRow={(rowData) =>
                        <TouchableOpacity onPress={() => {
                            this.setState({selectedSource: rowData.url, dataSource: this.ds.cloneWithRows(this.urls)});
                            ReactNativeAudioStreaming.play(rowData.url, {});
                        }}>
                            <View style={StyleSheet.flatten([
                                styles.row,
                                {backgroundColor: rowData.url == this.state.selectedSource ? '#3fb5ff' : 'white'}
                            ])}>
                                <Text style={styles.icon}>▸</Text>
                                <View style={styles.column}>
                                    <Text style={styles.name}>{rowData.name}</Text>
                                    <Text style={styles.url}>{rowData.url}</Text>
                                </View>
                            </View>
                        </TouchableOpacity>
                    }
                />

                <Player url={this.state.selectedSource} />
            </View>
        );
    }
}

const styles = StyleSheet.create({
    container: {
        flex: 1,
        justifyContent: 'center',
        alignItems: 'center',
        backgroundColor: '#F5FCFF',
        paddingTop: Platform.OS === 'ios' ? 30 : 0
    },
    row: {
        flex: 1,
        flexDirection: 'row',
        padding: 5,
        borderBottomColor: 'grey',
        borderBottomWidth: 1
    },
    column: {
        flexDirection: 'column'
    },
    icon: {
        fontSize: 26,
        width: 30,
        textAlign: 'center'
    },
    name: {
        color: '#000'
    },
    url: {
        color: '#CCC'
    }
});
